[AMDGPU] GCNRegPressure printing pass for testing.

vpykhtin opened the pull request:
Take a look at the tests to see the output.

This probably should be also enabled for _llc_ tests, I haven't tried it yet.

vpykhtin committed:
[AMDGPU] GCNRegPressure printing pass for testing.

vpykhtin committed:
Various improvents:
 * Added printing of live-in, live-out and live-through sets.
 * Empty BB aren't skipped now.

rampitec commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -487,3 +488,77 @@ LLVM_DUMP_METHOD
+
+  const MachineRegisterInfo &MRI = MF.getRegInfo();
+  const LiveIntervals &LIS = getAnalysis<LiveIntervals>();
+  GCNUpwardRPTracker RPT(LIS);
```
I thought we are using GCNDownwardRPTracker in the scheduling.

perlfu replied:
What kind of difference are we expecting between the two?
Would it makes sense if it was configurable?

vpykhtin replied:
Sorry I had to provide more context about this.
 
Piotr has found the problem with upward tracker [D33289](https://reviews.llvm.org/D33289) and Jay suggested to make a debug pass to test our RP values before fixing the tracker.

I think we should test both trackers and in theory they should give the same results so we can do an upward and downward runs and check their output with a single Filecheck prefix. If they don't match test update tool will warn about this and we can add a second and third prefixes to see the difference.

I've faced a situation when the trackers disagree, let me add the downward tracker option to this pass and I'll try to illustrate how it works.

Very briefly:
```
  undef %0.sub1:sgpr_64 = ...
  use %0
```
LiveIntervals creates only a subrange for _sub1_ subreg but not a subrange for undefined _sub0_, but _sub0_ is used when %0 is accessed.

vpykhtin replied:
They disagree from the very first attempt and it is very inconvenient to have two Filecheck prefixes for upward and downward trackers because it's not easy to compare. We can track RP both ways and print the difference at each position if it exists.

perlfu commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -487,3 +488,77 @@ LLVM_DUMP_METHOD
+    MBB.printName(OS);
+    OS << ":\n";
+
+    if (MBB.empty()) {
```
I think this could be tweaked to avoid duplication of the printing logic.
It would also be helpful to have live-through output for all blocks, even if this can be inferred from live-in versus live-out, it is helpful to human readers.

vpykhtin committed:
Added:
 * downward tracker
 * skip dbg values

vpykhtin commented:
I've added downward tracker as a separate pass to show how it looks.

arsenm commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -487,3 +488,111 @@ LLVM_DUMP_METHOD
+INITIALIZE_PASS(GCNRegPressurePrinter, "amdgpu-print-rp", "", true, true)
+
+bool GCNRegPressurePrinter::runOnMachineFunction(MachineFunction &MF) {
+  if (skipFunction(MF.getFunction()))
```
A printer wouldn't ever skip a function?

vpykhtin replied:
removed skipping.

arsenm commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -487,3 +488,111 @@ LLVM_DUMP_METHOD
+  auto &OS = dbgs();
+
+// Leading spaces are important for YAML syntax.
+#define PFX "  "
```
does a constexpr constant string work?

vpykhtin replied:
unfortunately no, I use PFX in concatenation with other string literals.

arsenm commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -487,3 +488,111 @@ LLVM_DUMP_METHOD
+    
+    if (UseDownwardTracker) {
+      GCNDownwardRPTracker RPT(LIS);
+      RPT.reset(MBB.instr_front());
```
why instr_front instead of just front?

vpykhtin replied:
replaced with front.

vpykhtin committed:
restore tests

vpykhtin committed:
Improved version + comments.

vpykhtin commented:
Improved version:

* added live-through registers sets.
* added check if tracked live register set matches the set reported by LiveIntervals.
* debug values on BB boundaries are correctly skipped now.
* minimized special processing of empty basic blocks (I had to modify upward tracker interface).
* added testcase to illustrate LiveIntervals problem with undefined subregs,
  please take a look at upward_problem_lis_subregs_mismatch testcase.
* other review issues solved.

I decided to leave separate passes for upward and downward trackers.

This is close to the final version but I need to try running this pass with llc lit tests.

vpykhtin committed:
* Fixed release build by removing NDEBUG guards around printing routines.
* Fixed formatting.

vpykhtin commented:
> but I need to try running this pass with llc lit tests.

I believe my printing pass is inapplicable for _.ll_ tests because I cannot run separate passes there?

I think I'm done with this patch.

P.S. If I understood correctly _-print<...>_ syntax is only available with the new pass manager which isn't available for our target yet?

jayfoad commented:
> P.S. If I understood correctly -print<...> syntax is only available with the new pass manager which isn't available for our target yet?

Yeah, it does not support MachineIR passes (for any target) yet.

jayfoad commented on llvm/test/CodeGen/AMDGPU/regpressure_printer.mir:
```diff
@@ -0,0 +1,388 @@
+  ; RPU-NEXT:   0     7
+  ; RPU-NEXT:   0     8      %4.sub0:vreg_64 = V_MOV_B32_e32 111, implicit $exec
+  ; RPU-NEXT:   0     8
+  ; RPU-NEXT:   0     10     %5:vreg_64 = COPY %2:vreg_64
```
RPU and RPD are both wrong here. Pressure "at" this instruction should be 9.

jayfoad commented on llvm/test/CodeGen/AMDGPU/regpressure_printer.mir:
```diff
@@ -0,0 +1,388 @@
+  ; RPD-NEXT:   0     8
+  ; RPD-NEXT:   0     10     %5:vreg_64 = COPY %2:vreg_64
+  ; RPD-NEXT:   0     9
+  ; RPD-NEXT:   0     10     undef %6.sub0:vreg_64 = V_ADD_F32_e32 %1.sub0:vreg_64, %5.sub0:vreg_64, implicit $mode, implicit $exec
```
RPD seems particularly wrong here. Pressure does not go as high as 10 here.

vpykhtin commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -483,7 +496,125 @@ Printable llvm::print(const GCNRPTracker::LiveRegSet &LiveRegs,
+    if (UseDownwardTracker)
+      ReportLISMismatchIfAny(LRAtMBBEnd, getLiveRegs(MBBEndSlot, LIS, MRI));
+
+    GCNRPTracker::LiveRegSet LRThr = getIntersection(LRAtMBBBegin, LRAtMBBEnd);
+    OS << PFX "  Live-thr:" << llvm::print(LRThr, MRI);
```
I thought "live through" means that a register isn't redefined in the block. However I forgot that we can have redefinitions after register coalescer and using only intersection isn't enough to tell if a register keeps its value through the block. I can use the result of intersection to check if a register is redefined (or used?)

vpykhtin replied:
Sorry, I mean using registers from the intersection I need to perform additional pass on def/used for a given block.

rampitec replied:
IMHO it shall mean 'no defs or kills in the block'. Uses itself are included.

vpykhtin replied:
Should be enough to test whether an MBB is in one liverange segment of reg or subreg.

vpykhtin replied:
Fixed live-through register set calculation, assuming live-through register is in live-in and live-out sets, has no redefinitions  but may have uses in the block.

vpykhtin commented on llvm/test/CodeGen/AMDGPU/regpressure_printer.mir:
```diff
@@ -0,0 +1,388 @@
+    S_NOP 0, implicit %0, implicit %1
+...
+
+# This testcase shows the problem with LiveIntervals: it doesn't create
+# subranges for undefined but used subregisters. Upward tracker is able to see
+# the use of undefined subregister and tracks it correctly.
```
But it cannot allocate half of register if it's required by an instruction?

jayfoad replied:
Between the first V_MOV and the last S_NOP, only half of %0:vreg_64 is live so the allocator only needs to dedicate one VGPR to it. All other VGPRs can be reused for other things, so they should not contribute to the register pressure.

vpykhtin replied:
Right, given that both trackers are already optimistic in that they believe the allocator can reuse for example, unused subregs of %0.sub0:sgpr_128 use, so this is the same situation here.

vpykhtin committed:
Fixed live-throught register set calculation, added a test.

vpykhtin commented:
I'm really sorry for spam here, wrong move within VSCode.

vpykhtin committed:
Merge branch 'main' into rp_printer

vpykhtin commented:
Should we move on and submit this patch? @jayfoad do you have concerns about live-through register set computation or others? I believe fixing trackers should go to another PR.

jayfoad reviewed (approved):
> Should we move on and submit this patch?

Yes!

> @jayfoad do you have concerns about live-through register set computation or others?

I personally have no interest in the live-through part. You could remove it from this patch, but I don't mind if others want to keep it.

> I believe fixing trackers should go to another PR.

Agreed.

vpykhtin committed:
Move live-through register set printing to another patch.

vpykhtin committed:
Merge branch 'main' into rp_printer

vpykhtin commented:
I've removed live-through registers printing from this PR and will submit it separately.
